src/Components/ElementsPanel/Internal/VisualRuleElementsList.tsx

```diff
@@ -0,0 +1,274 @@
+} from '../ViewerElementsPanel.types';
+import { sortPanelItemsForDisplay } from '../ViewerElementsPanel.Utils';
+import { useId } from '@fluentui/react-hooks';
+import { useExtendedTheme } from '../../../Models/Hooks/useExtendedTheme';
+
+const VisualRuleElementsList: React.FC<IViewerElementsPanelListProps> = ({
+    isLoading,
+    isModal,
+    panelItems = [],
+    filterTerm,
+    onItemClick,
+    onItemHover,
+    onItemBlur
+}) => {
+    const { t } = useTranslation();
+    const elementsPanelStyles = getElementsPanelStyles();
+
+    // handle only showing "Loading..." on first fetch
+    const isInitialDataLoaded = useRef(false);
+    useEffect(() => {
+        if (panelItems.length) {
+            isInitialDataLoaded.current = true;
+        }
+    }, [panelItems]);
+
+    const listItems = useMemo(
+        () =>
+            getListItems(
+                panelItems,
+                isModal,
+                onItemClick,
+                onItemHover,
+                onItemBlur
+            ),
+        [panelItems, onItemClick, onItemHover, onItemBlur]
+    );
+
+    return (
+        <div className={elementsPanelStyles.list}>
+            {isLoading && !isInitialDataLoaded.current ? (
+                <p style={{ padding: '0px 20px' }}>{t('loading')}</p>
+            ) : panelItems.length === 0 ? (
+                <p style={{ padding: '0px 20px' }}>
+                    {t('elementsPanel.noElements')}
+                </p>
+            ) : (
+                <CardboardList<ITwinToObjectMapping | IVisual>
+                    items={listItems}
+                    listKey={`elements-panel`}
+                    textToHighlight={filterTerm}
+                />
+            )}
+        </div>
+    );
+};
+
+function getListItems(
+    panelItems: Array<IViewerElementsPanelItem>,
+    isModal: boolean,
+    onItemClick: ElementsPanelCallback,
+    onItemHover?: ElementsPanelCallback,
+    onItemBlur?: ElementsPanelCallback
+): Array<ICardboardGroupedListItem<ITwinToObjectMapping | IVisual>> {
+    const sortedPanelItems = sortPanelItemsForDisplay(panelItems);
+    const buttonStyles = getElementsPanelButtonSyles();
+    const listItems: Array<
+        ICardboardGroupedListItem<ITwinToObjectMapping | IVisual>
+    > = [];
+
+    sortedPanelItems.map((panelItem) => {
+        const element = panelItem.element;
+        const elementColorings: Array<VisualColorings> = [];
+        const badges: Array<{
+            behavior: IBehavior;
+            visual: IValueRangeVisual;
+            visualRule: IExpressionRangeVisual;
+            visualRuleDisplayTitle: string;
+        }> = [];
+        let visualRules: IExpressionRangeVisual[] = [];
+
+        panelItem.behaviors.map((b) => {
+            // Add all visual rules to an array
+            visualRules = b.visuals.filter(ViewerConfigUtility.isVisualRule);
+
+            // Separate into their respective arrays
+            visualRules.forEach((vr) => {
```

Comment:
sooooo many nested loops here. This is going to get realllly bogged down on large scenes.